google/cloud/storage/internal/xml_parser.cc

```diff
@@ -0,0 +1,52 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/internal/xml_parser.h"
+#include "google/cloud/internal/absl_str_cat_quiet.h"
+
+namespace google {
+namespace cloud {
+namespace storage_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+std::string XmlParser::CleanUpXml(std::string const& content) {
+  auto ret = std::regex_replace(content, xml_decl_re_, "");
+  ret = std::regex_replace(ret, xml_doctype_re_, "");
+  ret = std::regex_replace(ret, xml_cdata_re_, "");
+  ret = std::regex_replace(ret, xml_comment_re_, "");
+  return ret;
+}
+
+StatusOr<std::shared_ptr<XmlNode>> XmlParser::ParseXml(
+    std::string const& content, Options options) {
+  internal::CheckExpectedOptions<XmlParserOptionsList>(options, __func__);
+  options = XmlParserDefaultOptions(std::move(options));
+
+  // Check size first
+  if (content.length() > options.get<XmlParserSourceMaxBytes>()) {
+    return internal::InvalidArgumentError(
+        absl::StrCat("XML exceeds the max byte size of ",
+                     options.get<XmlParserSourceMaxBytes>()),
+        GCP_ERROR_INFO());
+  }
+  // Remove unnecessary bits
+  auto trimmed = CleanUpXml(content);
```

Comment:
Can't we just skip over them during parsing rather than having to edit them out up front?

Reply:
Technically possible, but removing makes the parsing part much simpler.

Reply:
It is difficult to say without seeing the implementation of the parser.  At the moment all I can say is that copying the string four times sounds like a significant pessimization.  Might we at least create a single, disjunction RE that can do all four removals in one copy?

Reply:
It looks like it wouldn't be too complicated to skip during parsing, but OK.